src/util/RtreeBuild.cpp

```diff
@@ -0,0 +1,1133 @@
+#include <util/Rtree.h>
+#include <util/RtreeFileReader.h>
+#include <util/BackgroundStxxlSorter.h>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/serialization/vector.hpp>
+
+static void centerOrdering(multiBoxGeo& boxes, size_t dim) {
+  if (dim == 0) {
+    // order by centerX
+    SortRuleLambdaX comp;
+
+    std::sort(boxes.begin(), boxes.end(), comp);
+  } else {
+    // order by centerY
+    auto sortRuleLambda = [](RTreeValue b1, RTreeValue b2) -> bool {
+      double center1 = std::midpoint(b1.box.min_corner().get<1>(), b1.box.max_corner().get<1>());
+      double center2 = std::midpoint(b2.box.min_corner().get<1>(), b2.box.max_corner().get<1>());
+      return center1 < center2;
+    };
+
+    std::sort(boxes.begin(), boxes.end(), sortRuleLambda);
+  }
+}
+
+static void centerOrdering(multiBoxWithOrderIndex& boxes, size_t dim) {
+  if (dim == 0) {
+    // order by centerX
+    SortRuleLambdaXWithIndex comp;
+
+    std::sort(boxes.begin(), boxes.end(), comp);
+  } else {
+    // order by centerY
+    SortRuleLambdaYWithIndex comp;
+
+    std::sort(boxes.begin(), boxes.end(), comp);
+  }
+}
+
+OrderedBoxes SortInput(const std::string& onDiskBase, size_t M, uintmax_t maxBuildingRamUsage, bool workInRam) {
+  OrderedBoxes orderedInputRectangles;
+
+  ad_utility::BackgroundStxxlSorter<RTreeValue, SortRuleLambdaX> sorterRectsD0Basic = ad_utility::BackgroundStxxlSorter<RTreeValue, SortRuleLambdaX>(std::ceil((maxBuildingRamUsage < 9999999999 ? maxBuildingRamUsage : 9999999999) / 3.0));
+  multiBoxGeo rectsD0Basic;
+
+  if (workInRam) {
+    rectsD0Basic = Rtree::LoadEntries(onDiskBase + ".boundingbox.tmp");
+    centerOrdering(rectsD0Basic, 0);
+  } else {
+    FileReaderWithoutIndex fileReaderRectsD0 = FileReaderWithoutIndex(onDiskBase + ".boundingbox.tmp");
+    std::optional<RTreeValue> rectD0Element = fileReaderRectsD0.GetNextElement();
+    while (rectD0Element) {
+      sorterRectsD0Basic.push(rectD0Element.value());
+      rectD0Element = fileReaderRectsD0.GetNextElement();
+    }
+    fileReaderRectsD0.Close();
+  }
+
+  uint64_t xSize = 0;
+  double globalMinX = -1;
+  double globalMinY = -1;
+  double globalMaxX = -1;
+  double globalMaxY = -1;
+
+  ad_utility::BackgroundStxxlSorter<RTreeValueWithOrderIndex, SortRuleLambdaYWithIndex> sorterRectsD1 = ad_utility::BackgroundStxxlSorter<RTreeValueWithOrderIndex, SortRuleLambdaYWithIndex>(std::ceil((maxBuildingRamUsage < 9999999999 ? maxBuildingRamUsage : 9999999999) / 3.0));
+  std::shared_ptr<multiBoxWithOrderIndex> RectanglesD1WithOrder = std::make_shared<multiBoxWithOrderIndex>();
+
+  if (workInRam) {
+    for (RTreeValue element : rectsD0Basic) {
+      RTreeValueWithOrderIndex entry = {element.box, element.id, xSize, 0};
+      RectanglesD1WithOrder->push_back(entry);
+      xSize++;
+
+      if (globalMinX == -1 || element.box.min_corner().get<0>() < globalMinX) {
+        globalMinX = element.box.min_corner().get<0>();
+      }
```

Comment:
Can the (with the above initialization) be `globalMinx = std::min(globalMinX, element.minX())`
(Of course the `element` should implement `minX(), `maxX()` etc. to get rid of the `.box.min_corner().get<0>` that is all across your codebase.